File: Python/6py 2/valid_braces.py
"""
|--------------------------------------------------------------------------
| Valid Braces
|--------------------------------------------------------------------------
|
"""

def valid_braces(string):
   res, pairs = [], {"(": ")", "[": "]", "{": "}"}
   for x in string:
      if x in pairs: res.append(x)
      else: 
         if not res or pairs[res.pop()] != x: return False
   return len(res) == 0


def validBraces(s, previous = ''):
   while s != previous: previous, s = s, s.replace('[]','').replace('{}','').replace('()','')
   return not s


def validBraces(s):
   pairs = ['{}', '()', '[]']
   while any(pair in s for pair in pairs):
      for pair in pairs: 
         s = s.replace(pair, "")
   return s == ""


print(valid_braces("(){}[]"), True)
print(valid_braces("([{}])"), True)
print(valid_braces("(}"), False)
print(valid_braces("[(])"), False)
print(valid_braces("[({})](]"), False)

File: JavaScript/6js 3/validBraces.js
/*
|--------------------------------------------------------------------------
| Valid Braces
|--------------------------------------------------------------------------
|
*/

function validBraces(braces) {
   const res = [];
   const pairs = {
      "(": ")",
      "[": "]",
      "{": "}",
   };
   for (let i = 0; i < braces.length; i++) {
      if (braces[i] === "(" || braces[i] === "[" || braces[i] === "{") {
         res.push(braces[i]);
      } else {
         if (pairs[res.pop()] !== braces[i]) return false;
      }
   }
   return res.length === 0;
}


function validBraces(braces) {
   while (/\(\)|\[\]|\{\}/g.test(braces)) {
      braces = braces.replace(/\(\)|\[\]|\{\}/g, "");
   }
   return !braces.length;
}


function validBraces(braces) {
   for (var i = 0, depth = []; i < braces.length; i++) {
      switch (braces[i]) {
         case "(":
         case "[":
         case "{":
            depth.push(braces[i]);
            break;
         case ")":
            if (depth.pop() != "(") return false;
            break;
         case "]":
            if (depth.pop() != "[") return false;
            break;
         case "}":
            if (depth.pop() != "{") return false;
            break;
      }
   }
   return depth.length == 0;
}


console.log(validBraces("(){}[]"), true);
console.log(validBraces("([{}])"), true);
console.log(validBraces("(}"), false);
console.log(validBraces("[(])"), false);
console.log(validBraces("[({})](]"), false);